openlibrary/plugins/admin/code.py

```diff
@@ -249,6 +250,75 @@ def GET(self):
         pass
 
 
+class sync_ia_ol(delegate.page):
+    path = '/ia/sync'
+    encoding = 'json'
+
+    def POST(self):
+        i = web.input(digest="", msg="")
+
+        digest = i.digest
+        msg = i.msg
+
+        try:
+            HMACToken.verify(digest, msg, "ia_sync_secret")
+        except (ValueError, ExpiredTokenError):
+            raise web.HTTPError("401 Unauthorized")
+
+        op, ocaid, _ = msg.split("|")
+
+        if not op or not ocaid:
+            raise web.HTTPError(
+                "400 Bad Request", data=json.dumps({"error": "Invalid inputs"})
+            )
+
+        # Fetch affected editions
+        if not (
+            edition_keys := web.ctx.site.things(
+                {"type": '/type/edition', "ocaid": ocaid}
+            )
+        ):
+            raise web.HTTPError("404 Not Found")
+
+        editions = [web.ctx.site.get(key) for key in edition_keys]
+        if len(editions) > 1:
+            raise web.HTTPError(
+                "409 Conflict",
+                data=json.dumps(
+                    {"error": "Multiple editions associated with given ocaid"}
+                ),
+            )
+
+        edition = editions[0]
+
+        # Update records
+        try:
+            match op:
+                case "makedark":
+                    self.make_dark(edition)
+                case "updatemarc":
+                    # XXX : Where is the new marc record coming from?
+                    new_marc = None
+                    self.update_marc(edition, new_marc)
+                case _:
+                    raise ValueError(f"Unknown operation : {op}")
+        except (NotImplementedError, ValueError):
+            raise web.HTTPError(
+                "400 Bad Request",
+                data=json.dumps({"error": f"Unrecognized operation : {op}"}),
+            )
+
+        return delegate.RawText(json.dumps({"status": "ok"}))
+
+    def make_dark(self, edition):
+        data = edition.dict()
+        del data["ocaid"]
+        web.ctx.site.save(data, 'Remove OCAID: Item no longer available to borrow.')
```

Comment:
The MARC record should also be disassociated here.  These can be found in the edition's `source_records` list, and have an `ia:` prefix.